### ISSUE TITLE
Simple example run-all.sh fails on first run

### ISSUE THREAD

**Issue by KevinWMatthews** [interledger-rs/interledger-rs#202]
I'm seeing `run-all.sh` fail to send payments the very first time after the `interledger` library is rebuilt, for example after pulling the latest master or making a change to a crate.

My current theory (up for revision) is that `start-services.sh` builds all *binaries* but runs nodes as *packages*, so starting a node actually rebuilds the `interledger` package. This rebuild happens in the background, so the main script tries to send payments to non-existent nodes.

In detail: `start-services.sh` runs `cargo build --bins` run to completion:

```
Building interledger.rs... (this may take a little while)
   Compiling interledger v0.4.0 (/home/kevin/coding/ilp/interledger-rs/crates/interledger)
    Finished dev [unoptimized + debuginfo] target(s) in 41.13s
```

and `redis` start successfully.

After that, compilation happens again while the the nodes are starting up:

```
Blocking waiting for file lock on package cache lock
Blocking waiting for file lock on build directory
Compiling interledger v0.4.0 (/home/kevin/coding/ilp/interledger-rs/crates/interledger)
The Interledger.rs nodes are up and running!=======================> ] 301/303: interledger   
```

This is logging to the console per #201 so output is intermingled.

At this point, `run-all.sh` moves on and sends payment:

```
Creating accounts:

Alice's account:
curl: (7) Failed to connect to localhost port 7770: Connection refused

# snip

Bob's Account:
curl: (7) Failed to connect to localhost port 8770: Connection refused

# ...and so on
```

These fail because the nodes are still being compiled.

`run-all.sh` then terminates everything using `kill-services.sh`:

```
Stopping Interledger nodes
30325:M 09 Aug 09:29:26.084 # User requested shutdown...
30325:M 09 Aug 09:29:26.084 * Saving the final RDB snapshot before exiting.
30325:M 09 Aug 09:29:26.130 * DB saved on disk
30325:M 09 Aug 09:29:26.133 # Redis is now ready to exit, bye bye...
```

Some time later, output from node compilation shows up on the console:

```
kevin@kevin-LM ~/coding/ilp/interledger-rs/examples/simple (master) $     Building [=====================================================> ] 302/303: interledger(bin)
    Finished dev [unoptimized + debuginfo] target(s) in 38.90s
     Running `/home/kevin/coding/ilp/interledger-rs/target/debug/interledger node --config ./node-a/config.yml`
[2019-08-09T16:29:57Z DEBUG interledger::node] Starting Interledger node with ILP address: example.node-a
[2019-08-09T16:29:57Z DEBUG interledger_store_redis::store] Connected to redis: Client { connection_info: ConnectionInfo { addr: Tcp("127.0.0.1", 6379), db: 0, passwd: None } }
[2019-08-09T16:29:57Z ERROR interledger_store_redis::store] Error connecting to Redis: Connection refused (os error 111)
[2019-08-09T16:29:57Z ERROR interledger::node] Error connecting to Redis: Tcp("127.0.0.1", 6379) ()
    Finished dev [unoptimized + debuginfo] target(s) in 38.97s
     Running `/home/kevin/coding/ilp/interledger-rs/target/debug/interledger node --config ./node-b/config.yml`
[2019-08-09T16:29:57Z DEBUG interledger::node] Starting Interledger node with ILP address: example.node-b
[2019-08-09T16:29:57Z DEBUG interledger_store_redis::store] Connected to redis: Client { connection_info: ConnectionInfo { addr: Tcp("127.0.0.1", 6379), db: 1, passwd: None } }
[2019-08-09T16:29:57Z ERROR interledger_store_redis::store] Error connecting to Redis: Connection refused (os error 111)
[2019-08-09T16:29:57Z ERROR interledger::node] Error connecting to Redis: Tcp("127.0.0.1", 6379) ()
```

The second time around things are fine because both binaries and packages have been built.

I think that a solution is to either recompile the `interledger` package:

```
cargo build --package interledger
```

or run the nodes as binaries:

```
cargo run --bin interledger -- # etc
```

but I'm not sure at this point. Thoughts? Can anyone else reproduce this? I think this will still be present after #190 is merged.

**Comment by emschwartz** [comment]
Thanks for reporting this issue! Can you try using the script in #190? That runs each command sequentially so it should properly wait for the build to finish

**Comment by dora-gt** [comment]
Hi Kevin,

Please refer to [the WIP branch](https://github.com/dora-gt/interledger-rs/tree/feature/documentation/consistency-improvement/examples/simple#troubleshooting).

I think it is because the node is not ready yet (trying to start up and doesn't accept connections at that time) and please try changing [this](https://github.com/emschwartz/interledger-rs/blob/master/examples/simple/start-services.sh#L20) to 3 or larger.

**Comment by KevinWMatthews** [comment]
@emschwartz @dora-gt Sure, I'll try it on the new branch.

Looks like you're putting the script's executable code in a README and then running it using [run-md.sh](https://github.com/dora-gt/interledger-rs/blob/feature/documentation/consistency-improvement/scripts/run-md.sh). Cool idea. I'm using the [feature/documentation/consistency-improvement](https://github.com/dora-gt/interledger-rs/tree/feature/documentation/consistency-improvement/examples/simple) branch.

I still see the same issue.

Here's an excerpt from `node-b.log`:

```
    Blocking waiting for file lock on package cache lock
    Blocking waiting for file lock on package cache lock
   Compiling interledger-settlement v0.1.0 (/home/kevin/coding/ilp/interledger-rs/crates/interledger-settlement)
   Compiling interledger-service-util v0.2.1 (/home/kevin/coding/ilp/interledger-rs/crates/interledger-service-util)
   Compiling interledger-api v0.1.0 (/home/kevin/coding/ilp/interledger-rs/crates/interledger-api)
   Compiling interledger-store-memory v0.2.1 (/home/kevin/coding/ilp/interledger-rs/crates/interledger-store-memory)
   Compiling interledger-store-redis v0.2.1 (/home/kevin/coding/ilp/interledger-rs/crates/interledger-store-redis)
   Compiling interledger v0.4.0 (/home/kevin/coding/ilp/interledger-rs/crates/interledger)
    Finished dev [unoptimized + debuginfo] target(s) in 1m 32s
```

Even though `cargo build --bins` has finished, starting a node still recompiles to the tune of 90s (I'm on an old laptop).

The nodes are launched sequentially but they are run in the background:

```
cargo run --package interledger -- node &> logs/node-b.log &
```

so the nodes won't be running until well after the main script exits.

**Comment by KevinWMatthews** [comment]
Does this behavior meet your expectations?

Edit `crates/interledger/src/lib.rs` (say, add trivial whitespace).

Build, then run:

```
$ cargo build
   Compiling interledger v0.4.0 (/home/kevin/coding/ilp/interledger-rs/crates/interledger)
    Finished dev [unoptimized + debuginfo] target(s) in 1m 04s
$ cargo run --package interledger -- node -c config.yaml
   Compiling interledger v0.4.0 (/home/kevin/coding/ilp/interledger-rs/crates/interledger)
    Finished dev [unoptimized + debuginfo] target(s) in 1m 10s
     Running `/home/kevin/coding/ilp/interledger-rs/target/debug/interledger node -c config.yaml`
```

 (I created a `config.yaml` file for convenience). `cargo run` compiles, then runs.

I can circumvent the issue by running with `--bin` instead of `--package`:

```bash
# edit crates/interledger/src/lib.rs
$ cargo build
   Compiling interledger v0.4.0 (/home/kevin/coding/ilp/interledger-rs/crates/interledger)
    Finished dev [unoptimized + debuginfo] target(s) in 56.28s
$ cargo run --bin interledger -- node -c config.yaml
    Finished dev [unoptimized + debuginfo] target(s) in 2.87s
     Running `/home/kevin/coding/ilp/interledger-rs/target/debug/interledger node -c config.yaml`
```

Alternatively, I can both build and run with `--package`:

```bash
# crates/interledger/src/lib.rs
$ cargo build --package interledger
   Compiling interledger v0.4.0 (/home/kevin/coding/ilp/interledger-rs/crates/interledger)
    Finished dev [unoptimized + debuginfo] target(s) in 56.83s
$ cargo run --package interledger -- node -c config.yaml
    Finished dev [unoptimized + debuginfo] target(s) in 1.79s
     Running `/home/kevin/coding/ilp/interledger-rs/target/debug/interledger node -c config.yaml`
```

**Comment by dora-gt** [comment]
AFAIK `cargo run` rebuilds only if compilation is required and I just tried to check it, it didn't rebuild.

Anyway, the problem that your `curl` isn't accepted is because the server isn't ready yet at that time regardless of rebuilds. So I fixed the branch to wait for the servers to start up. Could you try it?

**Comment by KevinWMatthews** [comment]
> `cargo run` rebuilds only if compilation is required

That is my impression as well. I'm perplexed why I'm seeing it compile at all.


> curl isn't accepted is because the server isn't ready yet at that time regardless of rebuilds.

The logs below seem to indicate otherwise.


> So I fixed the branch to wait for the servers to start up.

The loop is a nice touch. It now runs the properly first time after waiting for `cargo run` to finish compiling.

I still see the same behavior: `cargo run` rebuilds the first time, doesn't the second.

I also have a second, more powerful PC. Here is the output from this new machine and a fresh clone:

```
$ cd examples/simple/
$ ../../scripts/run-md.sh README.md
Stopping Interledger nodes

Building interledger.rs... (This may take a couple of minutes)
   # ... Compilation output here
   Compiling interledger v0.4.0 (interledger-rs/crates/interledger)
   Compiling interledger-settlement-engines v0.1.0 (interledger-rs/crates/interledger-settlement-engines)
    Finished dev [unoptimized + debuginfo] target(s) in 3m 23s

Waiting for nodes to start up...
................................................
The Interledger.rs nodes are up and running!
```

It takes quite some time to start the nodes.

Looking at `node-b.log`, I see:

```
$ cat logs/node-b.log 
    Blocking waiting for file lock on package cache lock
    Blocking waiting for file lock on package cache lock
   Compiling arrayvec v0.4.11
   Compiling memchr v2.2.1
   # ...and so on
   Compiling redis v0.10.0
   Compiling tower-web v0.3.7
   Compiling interledger-spsp v0.2.1 (interledger-rs/crates/interledger-spsp)
   Compiling interledger-http v0.2.1 (interledger-rs/crates/interledger-http)
   Compiling interledger-settlement v0.1.0 (interledger-rs/crates/interledger-settlement)
   Compiling interledger-service-util v0.2.1 (interledger-rs/crates/interledger-service-util)
   Compiling interledger-api v0.1.0 (interledger-rs/crates/interledger-api)
   Compiling interledger-store-memory v0.2.1 (interledger-rs/crates/interledger-store-memory)
   Compiling interledger-store-redis v0.2.1 (interledger-rs/crates/interledger-store-redis)
   Compiling interledger v0.4.0 (interledger-rs/crates/interledger)
    Finished dev [unoptimized + debuginfo] target(s) in 47.31s
     Running `interledger-rs/target/debug/interledger node`
```

Running a second time, it starts quickly:

```
$ ../../scripts/run-md.sh README.md 
Stopping Interledger nodes

Building interledger.rs... (This may take a couple of minutes)
    Finished dev [unoptimized + debuginfo] target(s) in 0.71s

Waiting for nodes to start up...
..
The Interledger.rs nodes are up and running!
```

```
$ cat logs/node-b.log
    Blocking waiting for file lock on package cache lock
    Blocking waiting for file lock on package cache lock
    Finished dev [unoptimized + debuginfo] target(s) in 0.42s
     Running `interledger-rs/target/debug/interledger node`
```

Note that the compilation step is absent.

Could this be a tooling issue? Both of my PCs are running:

```
cargo 1.36.0 (c4fcfb725 2019-05-15)
rustc 1.36.0 (a53f9df32 2019-07-03)
```

**Comment by dora-gt** [comment]
@KevinWMatthews Can this be closed now? or do you still have problems on examples?

**Comment by KevinWMatthews** [comment]
@dora-gt Looks good so far! If I encounter something else I'll open a new issue.
